File: src/unix/ibus/config_util.cc
// Copyright 2010-2011, Google Inc.
// All rights reserved.
//
// Redistribution and use in source and binary forms, with or without
// modification, are permitted provided that the following conditions are
// met:
//
//     * Redistributions of source code must retain the above copyright
// notice, this list of conditions and the following disclaimer.
//     * Redistributions in binary form must reproduce the above
// copyright notice, this list of conditions and the following disclaimer
// in the documentation and/or other materials provided with the
// distribution.
//     * Neither the name of Google Inc. nor the names of its
// contributors may be used to endorse or promote products derived from
// this software without specific prior written permission.
//
// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
// "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
// LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
// A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
// OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
// LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
// DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
// THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

#include "unix/ibus/config_util.h"

#include "base/base.h"
#include "base/protobuf/descriptor.h"

namespace mozc {
namespace ibus {
#if IBUS_CHECK_VERSION(1, 3, 99)
// IBus-1.4 uses Glib's GVariant for configration.
bool ConfigUtil::GetString(GVariant *value, const gchar **out_string) {
  if (g_variant_classify(value) != G_VARIANT_CLASS_STRING) {
    return false;
  }
  *out_string = g_variant_get_string(value, NULL);
  return true;
}

bool ConfigUtil::GetInteger(GVariant *value, gint *out_integer) {
  if (g_variant_classify(value) != G_VARIANT_CLASS_INT32) {
    return false;
  }
  *out_integer = g_variant_get_int32(value);
  return true;
}

bool ConfigUtil::GetBoolean(GVariant *value, gboolean *out_boolean) {
  if (g_variant_classify(value) != G_VARIANT_CLASS_BOOLEAN) {
    return false;
  }
  *out_boolean = g_variant_get_boolean(value);
  return true;
}
#else
// IBus-1.2 and 1.3 use GValue for configration.
bool ConfigUtil::GetString(GValue *value, const gchar **out_string) {
  if (!G_VALUE_HOLDS_STRING(value)) {
    return false;
  }
  *out_string = g_value_get_string(value);
  return true;
}

bool ConfigUtil::GetInteger(GValue *value, gint *out_integer) {
  if (!G_VALUE_HOLDS_INT(value)) {
    return false;
  }
  *out_integer = g_value_get_int(value);
  return true;
}

bool ConfigUtil::GetBoolean(GValue *value, gboolean *out_boolean) {
  if (!G_VALUE_HOLDS_BOOLEAN(value)) {
    return false;
  }
  *out_boolean = g_value_get_boolean(value);
  return true;
}
#endif

void ConfigUtil::SetFieldForName(const gchar *name,
#if IBUS_CHECK_VERSION(1, 3, 99)
                                 GVariant *value,
#else
                                 GValue *value,
#endif
                                 protobuf::Message *result) {
  if (!name || !value) {
    LOG(ERROR) << "name or value is not specified";
    return;
  }
  DCHECK(result);

  const google::protobuf::Descriptor *descriptor = result->GetDescriptor();
  const google::protobuf::Reflection *reflection = result->GetReflection();
  const google::protobuf::FieldDescriptor *field_to_update =
      descriptor->FindFieldByName(name);

  if (!field_to_update) {
    LOG(ERROR) << "Unknown field name: " << name;
    return;
  }

  // Set |value| to |result|.
  switch (field_to_update->cpp_type()) {
    case google::protobuf::FieldDescriptor::CPPTYPE_ENUM: {
      // |value| should be STRING.
      const gchar *string_value = NULL;
      if (!GetString(value, &string_value)) {
        LOG(ERROR) << "Bad value type for " << name;
        return;
      }
      DCHECK(string_value);
      const google::protobuf::EnumValueDescriptor *enum_value =
          descriptor->FindEnumValueByName(string_value);
      if (!enum_value) {
        LOG(ERROR) << "Bad value for " << name << ": " << string_value;
        return;
      }
      reflection->SetEnum(result, field_to_update, enum_value);
      VLOG(2) << "setting field: " << name << " = " << string_value;
      break;
    }
    case google::protobuf::FieldDescriptor::CPPTYPE_UINT32: {
      // unsigned int is not supported as chrome's preference type and int is
      // used as an alternative type, so |value| should be INT.
      gint int_value = -1;
      if (!GetInteger(value, &int_value)) {
        LOG(ERROR) << "Bad value type for " << name;
        return;
      }
      reflection->SetUInt32(result, field_to_update, int_value);
      VLOG(2) << "setting field: " << name << " = " << int_value;
      break;
    }
    case google::protobuf::FieldDescriptor::CPPTYPE_BOOL: {
      // |value| should be BOOLEAN.
      gboolean boolean_value = FALSE;
      if (!GetBoolean(value, &boolean_value)) {
        LOG(ERROR) << "Bad value type for " << name;
        return;
      }
      reflection->SetBool(result, field_to_update, boolean_value);
      VLOG(2) << "setting field: " << name << " = "
              << (boolean_value ? "true" : "false");
      break;
    }
    default: {
      // TODO(yusukes): Support other types.
      LOG(ERROR) << "Unknown or unsupported type: " << name << ": "
                 << field_to_update->cpp_type();
      return;
    }
  }
}

}  // namespace ibus
}  // namespace mozc

File: src/android/src/com/google/android/inputmethod/japanese/ConfigUtil.java
// Copyright 2010-2013, Google Inc.
// All rights reserved.
//
// Redistribution and use in source and binary forms, with or without
// modification, are permitted provided that the following conditions are
// met:
//
//     * Redistributions of source code must retain the above copyright
// notice, this list of conditions and the following disclaimer.
//     * Redistributions in binary form must reproduce the above
// copyright notice, this list of conditions and the following disclaimer
// in the documentation and/or other materials provided with the
// distribution.
//     * Neither the name of Google Inc. nor the names of its
// contributors may be used to endorse or promote products derived from
// this software without specific prior written permission.
//
// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
// "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
// LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
// A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
// OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
// LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
// DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
// THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

package org.mozc.android.inputmethod.japanese;

import org.mozc.android.inputmethod.japanese.preference.PreferenceUtil;
import org.mozc.android.inputmethod.japanese.protobuf.ProtoConfig.Config;
import org.mozc.android.inputmethod.japanese.protobuf.ProtoConfig.Config.FundamentalCharacterForm;
import org.mozc.android.inputmethod.japanese.protobuf.ProtoConfig.Config.HistoryLearningLevel;
import org.mozc.android.inputmethod.japanese.protobuf.ProtoConfig.GeneralConfig;

import android.content.SharedPreferences;

/**
 * Utilities to convert values stored in SharedPreferences into a Config proto instance.
 *
 * Note: Currently the primary configuration data is stored in SharedPreferences, and mozc server
 *   just reflects the configuration (via setConfig/setImposedConfig).
 *   However, when we start to support sync-features, mozc server may overwrite its configuration
 *   by itself. So we need to re-design configuration related stuff to support sync-features.
 */
public class ConfigUtil {
  // Disallow instantiation.
  private ConfigUtil() {
  }

  /**
   * Converts {@code SharedPreferences} to {@code Config}.
   *
   * Now this method only supports:
   * - space_input_character
   * - use_kana_modifier_insensitive_conversion
   * - use_typing_correction
   * - history_learning_level
   * - incognito_mode, and
   * - general_config (please see also toGeneralConfig(SharedPreferences)).
   */
  static Config toConfig(SharedPreferences sharedPreferences) {
    Config.Builder builder = null;

    FundamentalCharacterForm spaceCharacterForm =
        getFundamentalCharacterFormFromBooleanKey(
            sharedPreferences, PreferenceUtil.PREF_SPACE_CHARACTER_FORM_KEY);
    if (spaceCharacterForm != null) {
      builder = maybeCreateConfigBuilder(builder).setSpaceCharacterForm(spaceCharacterForm);
    }


    Boolean kanaModifierInsensitiveConversion = getBoolean(
        sharedPreferences, PreferenceUtil.PREF_KANA_MODIFIER_INSENSITIVE_CONVERSION_KEY);
    if (kanaModifierInsensitiveConversion != null) {
      builder = maybeCreateConfigBuilder(builder)
          .setUseKanaModifierInsensitiveConversion(
              kanaModifierInsensitiveConversion.booleanValue());
    }

    Boolean typingCorrection = getBoolean(
        sharedPreferences, PreferenceUtil.PREF_TYPING_CORRECTION_KEY);
    if (typingCorrection != null) {
      builder = maybeCreateConfigBuilder(builder)
          .setUseTypingCorrection(typingCorrection.booleanValue());
    }

    HistoryLearningLevel historyLearningLevel = getHistoryLearningLevel(
        sharedPreferences, PreferenceUtil.PREF_DICTIONARY_PERSONALIZATION_KEY);
    if (historyLearningLevel != null) {
      builder = maybeCreateConfigBuilder(builder)
          .setHistoryLearningLevel(historyLearningLevel);
    }

    Boolean incognitoMode = getBoolean(
        sharedPreferences, PreferenceUtil.PREF_OTHER_INCOGNITO_MODE_KEY);
    if (incognitoMode != null) {
      builder = maybeCreateConfigBuilder(builder)
          .setIncognitoMode(incognitoMode.booleanValue());
    }

    GeneralConfig generalConfig = toGeneralConfig(sharedPreferences);
    if (generalConfig != GeneralConfig.getDefaultInstance()) {
      builder = maybeCreateConfigBuilder(builder)
          .setGeneralConfig(toGeneralConfig(sharedPreferences));
    }

    if (builder != null) {
      return builder.build();
    }

    // No corresponding preference is found, so return the default instance.
    return Config.getDefaultInstance();
  }

  /**
   * Converts {@code SharedPreferences} to {@code GeneralConfig}.
   *
   * Now this method only supports:
   * - upload_usage_stats.
   */
  static GeneralConfig toGeneralConfig(SharedPreferences sharedPreferences) {
    GeneralConfig.Builder builder = null;

    Boolean uploadUsageStats = getBoolean(
        sharedPreferences, PreferenceUtil.PREF_OTHER_USAGE_STATS_KEY);
    if (uploadUsageStats != null) {
      builder = maybeCreateGeneralConfigBuilder(builder)
          .setUploadUsageStats(uploadUsageStats.booleanValue());
    }

    if (builder != null) {
      return builder.build();
    }

    // No corresponding preference is found, so return the default instance.
    return GeneralConfig.getDefaultInstance();
  }

  /**
   * @return {@code Boolean} value if {@code sharedPreferences} contains the value
   *   corresponding to the {@code key}. Otherwise returns {@code null}.
   */
  private static Boolean getBoolean(SharedPreferences sharedPreferences, String key) {
    if (!sharedPreferences.contains(key)) {
      return null;
    }

    // Default value wouldn't be used in actual case, but it is required.
    return Boolean.valueOf(sharedPreferences.getBoolean(key, false));
  }

  /**
   * Returns {@code FundamentalCharacterForm} value gotten from {@code sharedPreferences}.
   *
   * Note that FundamentalCharacterForm has three states (INPUT_MODE, FULL_WIDTH and HALF_WIDTH)
   * but current UI has only two states (INPUT_MODE and HALF_WIDTH).
   * This is because FULL_WIDTH doesn't work well for our alphabetical mode.
   * Even if FULL_WIDTH is set to the config,
   * <ul>
   * <li>On Precomposition mode if T13N is half-ascii a half-space is inserted regardless of
   * {@link Config#getSpaceCharacterForm()}.
   * <li>On Composition mode space-key event is mapped to Convert operation.
   * And on MechaMozc Convert operation on half-ascii mode behaves as
   * "Commit and insert half-space".
   * </ul>
   * Thus the behavior of FULL_WIDTH equals to INPUT_MODE.
   * Surely we can change above behavior but nobody won't want to input full-space on alphabet mode
   * so keeping the current implementation is reasonable.
   *
   * @return {@code FundamentalCharacterForm} value if {@code sharedPreferences} contains the value
   *   corresponding to the {@code key} (HALF_WIDTH for true, INPUT_MODE for false).
   *   Otherwise returns {@code null}.
   * @throws IllegalArgumentException if {@code FundamentalCharacterForm} doesn't have
   */
  private static FundamentalCharacterForm getFundamentalCharacterFormFromBooleanKey(
      SharedPreferences sharedPreferences, String key) {
    if (!sharedPreferences.contains(key)) {
      return null;
    }
    return sharedPreferences.getBoolean(PreferenceUtil.PREF_SPACE_CHARACTER_FORM_KEY, false)
        ? FundamentalCharacterForm.FUNDAMENTAL_HALF_WIDTH
        : FundamentalCharacterForm.FUNDAMENTAL_INPUT_MODE;
  }

  /**
   * @return {@code HistoryLearningLevel} value if {@code sharedPreferences} contains the value
   *   corresponding to the {@code key}. Otherwise returns {@code null}.
   * @throws IllegalArgumentException if {@code HistoryLearningLevel} doesn't have
   */
  private static HistoryLearningLevel getHistoryLearningLevel(
      SharedPreferences sharedPreferences, String key) {
    if (!sharedPreferences.contains(key)) {
      return null;
    }

    // Default value wouldn't be used in actual case, but it is required.
    return HistoryLearningLevel.valueOf(
        sharedPreferences.getString(key, HistoryLearningLevel.DEFAULT_HISTORY.name()));
  }

  private static Config.Builder maybeCreateConfigBuilder(Config.Builder builder) {
    if (builder != null) {
      return builder;
    }
    return Config.newBuilder();
  }

  private static GeneralConfig.Builder maybeCreateGeneralConfigBuilder(
      GeneralConfig.Builder builder) {
    if (builder != null) {
      return builder;
    }
    return GeneralConfig.newBuilder();
  }
}
